File: app/src/main/java/com/nikola/jakshic/dagger/di/DbModule.java
package com.nikola.jakshic.dagger.di;

import android.arch.persistence.room.Room;
import android.content.Context;

import com.nikola.jakshic.dagger.data.local.CompetitiveDao;
import com.nikola.jakshic.dagger.data.local.DotaDatabase;
import com.nikola.jakshic.dagger.data.local.HeroDao;
import com.nikola.jakshic.dagger.data.local.LeaderboardDao;
import com.nikola.jakshic.dagger.data.local.MatchDao;
import com.nikola.jakshic.dagger.data.local.PeerDao;
import com.nikola.jakshic.dagger.data.local.PlayerDao;
import com.nikola.jakshic.dagger.data.local.SearchHistoryDao;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

@Module(includes = NetworkModule.class)
public class DbModule {

    @Provides
    @Singleton
    PlayerDao providePlayerDao(DotaDatabase db) {
        return db.playerDao();
    }

    @Provides
    @Singleton
    CompetitiveDao provideCompetitiveDao(DotaDatabase db) {
        return db.competitiveDao();
    }

    @Provides
    @Singleton
    SearchHistoryDao provideSearchHistoryDao(DotaDatabase db) {
        return db.searchHistoryDao();
    }

    @Provides
    @Singleton
    PeerDao providePeerDao(DotaDatabase db) {
        return db.peerDao();
    }

    @Provides
    @Singleton
    HeroDao provideHeroDao(DotaDatabase db) {
        return db.heroDao();
    }

    @Provides
    @Singleton
    MatchDao provideMatchDao(DotaDatabase db) {
        return db.matchDao();
    }

    @Provides
    @Singleton
    LeaderboardDao provideLeaderboardDao(DotaDatabase db) {
        return db.leaderboardDao();
    }

    @Provides
    @Singleton
    DotaDatabase provideDotaDatabase(Context context) {
        return Room.databaseBuilder(context, DotaDatabase.class, "dagger.db").build();
    }
}

File: app/src/main/java/com/nikola/jakshic/dagger/di/DbModule.kt
package com.nikola.jakshic.dagger.di

import android.content.Context
import androidx.sqlite.db.SupportSQLiteDatabase
import app.cash.sqldelight.db.SqlDriver
import app.cash.sqldelight.driver.android.AndroidSqliteDriver
import com.nikola.jakshic.dagger.Database
import com.nikola.jakshic.dagger.common.sqldelight.DaggerSchema
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.android.qualifiers.ApplicationContext
import dagger.hilt.components.SingletonComponent
import io.requery.android.database.sqlite.RequerySQLiteOpenHelperFactory
import javax.inject.Singleton

@Module
@InstallIn(SingletonComponent::class)
object DbModule {
    @Provides
    fun providePlayerBookmarkQueries(database: Database) = database.playerBookmarkQueries

    @Provides
    fun provideMatchBookmarkQueries(database: Database) = database.matchBookmarkQueries

    @Provides
    fun providePlayerStatsQueries(database: Database) = database.playerStatsQueries

    @Provides
    fun provideMatchStatsQueries(database: Database) = database.matchStatsQueries

    @Provides
    fun providePeerQueries(database: Database) = database.peerQueries

    @Provides
    fun provideHeroQueries(database: Database) = database.heroQueries

    @Provides
    fun provideMatchQueries(database: Database) = database.matchQueries

    @Provides
    fun provideSearchHistoryQueries(database: Database) = database.searchHistoryQueries

    @Provides
    fun providePlayerQueries(database: Database) = database.playerQueries

    @Provides
    fun provideCompetitiveQueries(database: Database) = database.competitiveQueries

    @Provides
    fun provideLeaderboardQueries(database: Database) = database.leaderboardQueries

    @Provides
    fun provideHeroAssetsQueries(database: Database) = database.heroAssetQueries

    @Provides
    @Singleton
    fun provideSqlDriver(@ApplicationContext context: Context): SqlDriver {
        return AndroidSqliteDriver(
            schema = DaggerSchema,
            context = context,
            name = "dagger.db",
            factory = RequerySQLiteOpenHelperFactory(),
            callback = object : AndroidSqliteDriver.Callback(DaggerSchema) {
                override fun onOpen(db: SupportSQLiteDatabase) {
                    db.execSQL("PRAGMA foreign_keys = ON;")
                }
            },
        )
    }

    @Provides
    @Singleton
    fun provideDatabase(driver: SqlDriver): Database {
        return Database(driver)
    }
}
